rfcs/0032-bazel.md

```diff
@@ -0,0 +1,427 @@
+
+### Terminology
+
+**Platform**: an execution environment. Usually synonymous with
+"machine", i.e. HW architecture plus OS, but in the context of Bazel a
+platform may include e.g. a toolchain.  For example, a machine with
+multiple toolchains installed may satisfy multiple platform
+specifications.
+
+**Host**:
+- **Bazel host**, **orchestration host**: the host on which Bazel runs,
+  orchestrating the build.
+- **Build host**, **construction host**: the host on which the build tools
+  proper (compiler, linker, etc.) run, constructing the output.
+- **Target host**: the host for which output is constructed.
+
+**Build type**:
+- **Local builds**: the same machine plays all three host roles. This is
+  the most common situation.
+- **Distributed builds**: Bazel supports remote execution, so
+  orchestration and construction could run on different
+  machines/platforms, targeting yet a third machine/platform. This
+  does not necessarily imply different platforms; the machines could
+  be identical, except that the construction machine might faster,
+  have more RAM, etc.
+- **Mono-builds**: for lack of a better term. Same platform for all three
+  host roles.  A "native" build is a local mono-build.
+- **Cross-platform builds**: a/k/a "cross builds". The Target host
+  platform is different than the Bazel and Build host platforms.  A
+  cross-platform build could be either local or distributed.
+- **Canadian builds**: all three hosts are distinct platforms.  E.g. you
+  run Bazel on a mac, orchstrating build tools that run on a remote
+  linux machine, targeting Windows.
+
+### "Just works"
+
+The sense of "just works" is clear enough intuitively, but exactly
+what things/tasks should "just work" is a policy matter. For example,
+native builds clearly must just work, but which platforms should be
+supported is a policy decision. Should Windows native builds just
+work? If so, should the Cygwin toolchain be supported?  And so forth.
+
+What counts as just working in general:
+
+- Default local mono-builds ("native" builds) succeed on all supported
+  platforms; i.e. running build without passing any options by command
+  line or config file. Default settings just work, but may not be
+  optimal in every case.
+- Build refinements are easy and just work.  It must be easy for the
+  developer to express desired properties, and the ensuing build
+  configuration must succeed.  For example, optimization or debugging
+  flags.
+- Supported cross-platform builds must be easily expressible.  For example, it
+  should be possible to target Android by something like `$ bazel
+  build --platform=android_armv7`.
+- Configuration settings must be easily expressible either by command
+  line or via a config file (`.bazelrc`, `--bazelrc=file`).
+
+#### Integrated testing
+
+Bazel has builtin support for test targets.
+
+#### Cross-platform builds
+
+- Linux target from Mac host is in scope.  This is needed in order to
+  create docker images on the mac.
```

Comment:
Why is this needed to create docker images on mac? For example, `make build` works from within the `codaprotocol/coda:toolchain-eb73e64d524c906384e9dbb23e9d0b8ae2be9e9e` Docker image from macOS.

Reply:
To be more precise: needed to create the linux binaries that go in the docker image.  I.e. in order to create the whole schmear on a mac.

Reply:
OTOH, presumably one could build the linux code in a linux docker image running on the mac, so I guess this should be downgraded from Required to Desirable but not Mandatory.